Feature blog for KEP-3333 Retroactive default StorageClass assignment (#37625)
kubernetes/website: content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

---
layout: blog

> title: "Retroactive default StorageClass"

Reviewer: ```suggestion
title: "Kubernetes 1.25: Retroactive default StorageClass"
```

date: 2022-11-11
slug: retroactive-default-storage-class
---

**Author:** Roman Bednář (Red Hat)

> In Kubernetes 1.25 a new Alpha feature was introduced to improve the way default StorageClass is assigned to a PersistentVolumeClaim (PVC). It is no longer required to create a default StorageClass first and PVC second in order to assign the class and any PVCs without StorageClass can be updated later. The feature will be Beta in Kubernetes 1.26.

Reviewer: ```suggestion
As an optional, alpha feature, the v1.25 release of Kubernetes introduced a change to the way
that a default StorageClass is assigned to a PersistentVolumeClaim (PVC). With that feature enabled,
you no longer need to create a default StorageClass first and PVC second in order to assign the
class; also any PVCs without a StorageClass assigned can be updated later.
The new behavior will become beta in Kubernetes 1.26.
```

Is this right? Also, is this on by default for v1.26?

Author: Yes, and it is on by default in 1.26: https://github.com/kubernetes/kubernetes/pull/113329

> See [Official Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment) for more details.

Reviewer: ```suggestion
You can read [retroactive default StorageClass assignment](/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment) in the Kubernetes documentation for more details about how to use that,
or you can read on to learn about why the Kubernetes project is making this change.
```

Reviewer 2: The above suggestion needs to be incorporated, please.

Author: > Also, would highly recommend that we use either nil or null even though they are the same and have the potential to confuse users.

Yes, it got a bit mixed up when applying suggestions. I believe @sftim mentioned we stick to `null` here: https://github.com/kubernetes/website/pull/37625#discussion_r1028495144

I've removed any references to `nil`.

## Why did StorageClass assignment need improvements

> When talking about PVCs and default StorageClasses you might think "It already works, when I create PVC it gets the default class just fine!" and that is correct because there already is a feature that causes admission controller to assign a class to **new** PVCs which is documented [here](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass).

Reviewer: ```suggestion
When considering PVCs and default StorageClasses you might think "It already works, when I create PVC it gets the default class just fine!" and that is correct because there already is an [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass)that assigns a StorageClass to **new** PVCs.
```

But what if there was no default StorageClass at the time of PVC creation? Users end up with a PVC that will never get a class and so no storage will be provisioned, the PVC is somewhat "stuck" at this point. From high level point of view there were two main use cases which might result in "stuck" PVCs and cause problems later down the road. Let's take a closer look at each of them.

### Changing default StorageClass did not have a good solution

   There are two options admins had when they wanted to change default StorageClass:

>    1. Creating the new StorageClass as default before removing the old one which results in having two defaults for a short period of time. If a user creates a PersistentVolumeClaim at this point with storageClassName set to `nil` (which means default StorageClass) Kubernetes 1.26+ will choose the newest default StorageClass and assign it to this PVC.

Reviewer: 1. Creating the new StorageClass as default before removing the old one which results in having two defaults for a short period of time. If a user creates a PersistentVolumeClaim at this point with storageClassName set to `null` (which means default StorageClass), Kubernetes 1.26+ will choose the newest default StorageClass and assign it to this PVC.

Author: 1. Do we want to mix `nil` and `null` terms? Won't that be confusing?

2. Above you mentioned avoiding statements like *"Kubernetes 1.26+ will..."* - while in this case this behavior is already [merged](https://github.com/kubernetes/kubernetes/pull/110559). I don't even have to mention the version here probably as it does not add any value - removing.

>    2. Remove the old default first and create new default StorageClass which results in having no default for a short period of time. If user creates a PersistentVolumeClaim with storageClassName set to `nil` (which means default StorageClass) the PVC will be in `Pending` state forever, then user has to fix this by deleting the PVC a creating it again once default StorageClass is available.

Reviewer 3: a creating it again -> and creating it again

>    2. Remove the old default first and create new default StorageClass which results in having no default for a short period of time. If user creates a PersistentVolumeClaim with storageClassName set to `nil` (which means default StorageClass) the PVC will be in `Pending` state forever, then user has to fix this by deleting the PVC a creating it again once default StorageClass is available.

Reviewer: These lines should not have leading whitespace. Also, please wrap the source code - it makes the work of localization teams much easier.

Author: Done. You mean wrap into `<code>` tag instead of ``` ` ``` right?

Looking at the scenarios above it is clear that neither of them is optimal.

### Resource ordering during cluster installation

If cluster installation tool needs to create resources that require storage (e.g. image registry) it was difficult to get the ordering right because any Pods that require storage and rely on default StorageClass presence will fail to create if there is no default at that point yet.

## What changed

> We've changed PersistentVolume (PV) controller in KCM, so it can assign default StorageClass to unbound PersistentVolumeClaims that have storageClassName set to `nil`. In order to allow the value change PersistentVolumeClaim admission in the API server also changed to allow changing the value from `nil` to actual StorageClass name.

Reviewer: ```suggestion
We've changed PersistentVolume (PV) controller so it can assign a default StorageClass to any unbound PersistentVolumeClaims that have a null (unset) `storageClassName`. In order to allow the value change PersistentVolumeClaim admission within the API server also changed to allow changing the value from unset
to an actual StorageClass name.
```

> ### `nil` vs `""` for storageClassName - does it matter?

Reviewer: In our API, we use JSON as the canonical serialization, and `nil` is not a valid JSON literal.
```suggestion
### Null `storageClassName` versus `storageClassName: ""` - does it matter? {#null-vs-empty-string}
```

All the following text needs to change too, to avoid mentioning `nil`. Although the only implementations of the Kubernetes API that fully pass conformance are Golang code, we don't expect end users to have any Golang coding ability.

Author: Done.

With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name. With these constraints in mind the feature changes the semantics of `nil` if default StorageClass is present to always mean "Give me a default" and `""` to mean "Give me PersistentVolume that also has `""` StorageClass name".

In other words we need to distinguish two main cases, either default StorageClass is not present in which case there's no behavior change or where there is a default in which case PVCs with `nil` won't bind to PVs anymore and instead will be updated to have the StorageClass.

The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.

| Without default class        | PVC storageClassName == `""` | PVC storageClassName == `nil` |
|------------------------------|------------------------------|-------------------------------|
| PV storageClassName == `""`  | binds                        | binds                         |
| PV without storageClassName  | binds                        | binds                         |

> | With default class          | PVC storageClassName == `""` | PVC storageClassName == `nil` |

Reviewer 3: Can you add a title for each of these two tables so we know which has the new feature enabled and which does not?

Reviewer: It might be better to use a custom HTML table where the first column uses `rowspan`, and have that in place of these two tables.

Alternatively, have the table in 2 sections, each section introduced by a cell spanning all columns.

Author: I've changed the table to HTML (first version you suggested). Not sure I got the idea right though.

|-----------------------------|------------------------------|-------------------------------|
| PV storageClassName == `""` | binds                        | class updates                 |
| PV without storageClassName | binds                        | class updates                 |

## How to use it

> You don't have to perform any configuration steps - the feature will work "out of the box" once it's GA.

Reviewer: Is the feature GA? If not, don't write "You don't have to perform any configuration steps" in present tense.

Author: Done.

> You don't have to perform any configuration steps - the feature will work "out of the box" once it's GA.

Reviewer: Is the feature GA? If not, don't write "You don't have to perform any configuration steps" in present tense.

Author: I wanted to make it clear that in Beta and GA it's already enabled and no configuration is needed. Maybe it's ok to leave just the sentence below that alpha testing needs feature gate and remove this one - the rest is understood.

Reviewer 4: Maybe

> If you want to test the feature in Alpha you need to enable FeatureGates in `kube-controller-manager` and `kube-apiserver`:

Reviewer: ```suggestion
If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate in the
kube-controller-manager and the kube-apiserver. Use the `--feature-gates` command line
argument:
```

```bash
--feature-gates="...,RetroactiveDefaultStorageClass=true"
```

### Test drive

If you would like to see the feature in action and verify it works fine in your cluster here's what you can try:

> 1. We'll use a very simple PersistentVolumeClaim.

Reviewer: ```suggestion
1. Define a basic PersistentVolumeClaim:
```

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:

>   name: csi-pvc

Reviewer: How do we know this will use a CSI driver?

Author: We don't in fact. I just had a cluster with only one csi driver present but surely it makes more sense to use a generic name here, changed it.

spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

2. Create the PersistentVolumeClaim when there is no default StorageClass. The PVC won't provision or bind (unless there is an existing, suitable PV already present) and will remain in `Pending` state.

> ```bash

Reviewer: This isn't a valid shell script.

Author: Ack, removing `bash`.

$ kc get pvc
NAME      STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
csi-pvc   Pending   
```

3. Configure one StorageClass as default.

> ```bash

Reviewer: This isn't a valid shell script.

Author: Ack, removing `bash`.

$ kc patch sc/csi-hostpath-sc -p '{"metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"}}}'
storageclass.storage.k8s.io/csi-hostpath-sc patched
```

> 4. Verify that PersistentVolumeClaims is now provisioned correctly and was updated retroactively with new default StorageClass.

Reviewer: ```suggestion
4. Verify that the PersistentVolumeClaim you defined was updated retroactively with the new
   default StorageClass, and now has a PersistentVolume that is provisioned correctly:
```

> ```bash

Reviewer: This isn't a valid shell script.

Reviewer 2: This needs to be addressed, please.

Author: Ack, removing `bash`.

$ kc get pvc
NAME      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
csi-pvc   Bound    pvc-06a964ca-f997-4780-8627-b5c3bf5a87d8   1Gi        RWO            csi-hostpath-sc   87m
```

### New metrics

> To help indicate that the feature is working as expected we also introduced new `retroactive_storageclass_total` metric to show how many times PV controller attempted to update PersistentVolumeClaim and `retroactive_storageclass_errors_total` to show how many of those attempts failed.

Reviewer: ```suggestion
To help you see that the feature is working as expected we also introduced a new `retroactive_storageclass_total` metric to show how many times that the PV controller attempted to update PersistentVolumeClaim, and `retroactive_storageclass_errors_total` to show how many of those attempts failed.
```

Author: Done.

## Getting involved

> If you would like to share feedback with us, you can do so on our [public Slack channel](https://app.slack.com/client/T09NY5SBT/C09QZFCE5).

Reviewer: Explain who "us" means. So far, you've made it sound like "us" would mean Red Hat as you haven't mentioned SIG Storage clearly enough for a reader to make the link.

Reviewer 2: This needs to be addressed, please.

---
layout: blog

> title: "Kubernetes 1.25: Retroactive default StorageClass"

Reviewer 3: This should be "Kubernetes 1.26"

Author: Changed.

> date: 2022-11-11

Reviewer 5: It should be 1"Kubernetes 1.26" and the date should be somewhere after 1.26  GA.  We want it as post-1.26 feature blog

Author: I see, the "Release blog published" date is 6th December 2022 so that date might make the most sense here.

slug: retroactive-default-storage-class
---

**Author:** Roman Bednář (Red Hat)

As an optional, alpha feature, the v1.25 release of Kubernetes introduced a change to the way that a default StorageClass is assigned to a PersistentVolumeClaim (PVC).
With that feature enabled, you no longer need to create a default StorageClass first and PVC second in order to assign the class; also any PVCs without a StorageClass assigned can be updated later.

> We expect to graduate this feature to beta in Kubernetes 1.26.

Author: Changed.

See [Official Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment) for more details.

## Why did StorageClass assignment need improvements

> When considering PVCs and default StorageClasses you might think "It already works, when I create PVC it gets the default class just fine!" and that is correct because there already is an [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass)that assigns a StorageClass to **new** PVCs.

Author: This should point out there's already a similar feature (admission controller) that assigns classes to new PVC but is not related to the new feature. 

Would something like this make sense?

```
Users might be already familiar with a similar feature that assigns default StorageClasses to **new** PVCs at the time of creation. This is currently handled by [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass).
```

> When considering PVCs and default StorageClasses you might think "It already works, when I create PVC it gets the default class just fine!" and that is correct because there already is an [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass)that assigns a StorageClass to **new** PVCs.

Reviewer 2: Just rephrased your sentence for grammatical correctness.
```suggestion
Users might already be familiar with a similar feature that assigns default StorageClasses to **new** PVCs at the time of creation. This is currently handled by the [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass).
```

### Resource ordering during cluster installation

If cluster installation tool needs to create resources that require storage (e.g. image registry) it was difficult to get the ordering right because any Pods that require storage and rely on default StorageClass presence will fail to create if there is no default at that point yet.

## What changed

We've changed PersistentVolume (PV) controller so it can assign a default StorageClass to any unbound PersistentVolumeClaims that have a null (unset) <code>storageClassName</code>.
In order to allow the value change PersistentVolumeClaim admission within the API server also changed to allow changing the value from unset to an actual StorageClass name.

### <code>nil</code> vs <code>""</code> for storageClassName - does it matter?

Before this feature was introduced those values were equal in terms of behavior and any PersistentVolumeClaim with <code>nil</code> or <code>""</code> storageClassName value would bind to existing PersistentVolume resource with storageClassName also set to <code>nil</code> or <code>""</code>.

With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name.

> With these constraints in mind the feature changes the semantics of <code>nil</code> if default StorageClass is present to always mean "Give me a default" and <code>""</code> to mean "Give me PersistentVolume that also has <code>""</code> StorageClass name". 

Reviewer 2: This sentence is confusing. I think breaking up the sentences as below would be of value.
```suggestion
With these constraints in mind, the feature changes the semantics of <code>nil</code>. If a default StorageClass is present, <code>nil<code> would translate to "Give me a default" and <code>""</code> would mean "Give me PersistentVolume that also has <code>""</code> StorageClass name." In the absence of a StorageClass, the behavior would remain unchanged.
```

> In other words we need to distinguish two main cases, either default StorageClass is not present in which case there's no behavior change or where there is a default in which case PVCs with <code>nil</code> won't bind to PVs anymore and instead will be updated to have the StorageClass.

Reviewer 2: Does this sound accurate because this is what I understood?
```suggestion
Summarizing the above, we've changed the semantics of <code>nil<code> so that its behavior depends on the presence or absence of a definition of default StorageClass.
```

---
layout: blog
title: "Kubernetes 1.26: Retroactive Default StorageClass"

> date: 2022-12-06

Reviewer 2: Do we have a date set? Or was this supposed to go out yesterday? Please amend the dates here per the 1.26 sheet.

Author: Wasn't sure which date is this supposed to represent, changing this to "Release blog published" date.

Author: I've also changed the name of the file to match this one.

---
layout: blog
title: "Kubernetes 1.26: Retroactive Default StorageClass"

> date: 2022-12-08

Reviewer: ```suggestion
date: 2023-01-05
```
and please update the path `content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md` to match.

We've also modified the PersistentVolumeClaim admission within the API server to allow the change of values from unset value to an actual StorageClass name.

### <code>nil</code> vs <code>""</code> for storageClassName - does it matter?

Before this feature was introduced, those values were equal in terms of behavior. Any PersistentVolumeClaim with the storageClassName set to <code>nil</code> or <code>""</code> would bind to an existing PersistentVolume resource with storageClassName also set to <code>nil</code> or <code>""</code>.

With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name.
With these constraints in mind, the feature changes the semantics of <code>nil</code>. If a default StorageClass is present, <code>nil<code> would translate to "Give me a default" and <code>""</code> would mean "Give me PersistentVolume that also has <code>""</code> StorageClass name." In the absence of a StorageClass, the behavior would remain unchanged.

Summarizing the above, we've changed the semantics of <code>nil<code> so that its behavior depends on the presence or absence of a definition of default StorageClass.

> The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.

Reviewer: (nit)
```suggestion
The tables below show all these cases to better describe when PVC binds and when its StorageClass gets updated.
```

Author: Fixed.

> <table>

Reviewer: Consider adding a `<caption>...</caption>` element (helps with accessibility)

Author: Added.

> <table>

Reviewer: I'd use a single “=“ if not writing it inside `<code>...</code>`; also, I'd put `storageClassName` in monospace (eg `<tt>storageClassName</tt>`).

Author: Done.

  <thead>
     <tr>
        <th colspan="2"></th>
        <th>PVC storageClassName == <code>""</code></th>
        <th>PVC storageClassName == <code>nil</code></th>
     </tr>
  </thead>
  <tbody>
     <tr>
        <td rowspan="2">Without default class</td>
        <td>PV storageClassName == <code>""</code></td>
        <td>binds</td>
        <td>binds</td>
     </tr>
     <tr>
        <td>PV without storageClassName</td>
        <td>binds</td>
        <td>binds</td>
     </tr>
     <tr>
        <td rowspan="2">With default class</td>
        <td>PV storageClassName == <code>""</code></td>
        <td>binds</td>
        <td>class updates</td>
     </tr>
     <tr>
        <td>PV without storageClassName</td>
        <td>binds</td>
        <td>class updates</td>
     </tr>
  </tbody>
</table>

## How to use it

> If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate in the kube-controller-manager and the kube-apiserver. Use the <code>--feature-gates</code> command line argument:

Reviewer: ```suggestion
If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate
in the kube-controller-manager and the kube-apiserver.
Use the `--feature-gates` command line argument:
```
Main thing here is to avoid HTML where Markdown would work.

Author: Fixed.

```bash
--feature-gates="...,RetroactiveDefaultStorageClass=true"
```

### Test drive

If you would like to see the feature in action and verify it works fine in your cluster here's what you can try:

1. Define a basic PersistentVolumeClaim:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-1
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

2. Create the PersistentVolumeClaim when there is no default StorageClass. The PVC won't provision or bind (unless there is an existing, suitable PV already present) and will remain in <code>Pending</code> state.
```
$ kc get pvc
NAME      STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
csi-pvc   Pending   
```

3. Configure one StorageClass as default.
```
$ kc patch sc/csi-hostpath-sc -p '{"metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"}}}'
storageclass.storage.k8s.io/csi-hostpath-sc patched

> ```

Reviewer: Whenever we suggest `hostPath` as an example, we *need* to point out that it's a terrible idea to make that storage driver available unrestricted in production clusters.

Either:
- make up a fictional CSI driver and use that instead
- clearly add text to warn readers

Author: Good point. Just changing the SC name to something generic like `my-storageclass` should be fine and we can avoid mentioning `hostPath` anywhere.
